src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -823,13 +823,46 @@ trait Infer extends Checkable {
         case _                                                              => onRight
       }
     }
+
     private def isAsSpecificValueType(tpe1: Type, tpe2: Type, undef1: List[Symbol], undef2: List[Symbol]): Boolean = tpe1 match {
       case PolyType(tparams1, rtpe1) =>
         isAsSpecificValueType(rtpe1, tpe2, undef1 ::: tparams1, undef2)
       case _                         =>
         tpe2 match {
           case PolyType(tparams2, rtpe2) => isAsSpecificValueType(tpe1, rtpe2, undef1, undef2 ::: tparams2)
-          case _                         => existentialAbstraction(undef1, tpe1) <:< existentialAbstraction(undef2, tpe2)
+          case _                         =>
+
+            if (settings.YdottySpecificity) {
+              // Backport of fix for https://github.com/scala/bug/issues/2509
+              // from Dotty https://github.com/lampepfl/dotty/commit/89540268e6c49fb92b9ca61249e46bb59981bf5a
+              val flip = new TypeMap(trackVariance = true) {
+                var subst: Map[Symbol, Symbol] = Map.empty
+                def apply(tp: Type): Type = tp match {
+                  case TypeRef(pre, sym, args) if variance > 0 && sym.typeParams.exists(_.isContravariant) =>
+                    val flippedSym =
+                      subst.get(sym) match {
+                        case None =>
+                          val flippedClassSym = sym.owner.newClassSymbol(sym.name.toTypeName, sym.pos, sym.flags)
```

Comment:
I don't think that's quite what's going on in Dotty's implementation. In the `flip` in Dotty the `TypeMap` traversal appears to traverse through to the type members which correspond to the type params of the `TypeRef` in the scala implementation. These type members then have their variance individually flipped, via a type alias, as you say. But I think the mechanism in scalac has to be different ... we have to work at the `TypeRef`/type params level.

On your first point, yes you're right that new symbols will be unrelated by `<:<` except in the trivial case where the outer type constructor is the same in both cases (here the `subst` map ensures that the same flipped symbol is used). I'm fairly confident that my implementation here matches the one in Dotty, but it's possible that the Dotty implementation gets this wrong as well.

I think however, that in cases where the outer type constructor differs, we will always want to say that neither is specific as the other. The following is the main problem case (included in the latest commit, Dotty agrees with the implementation here),
```scala
trait Z[-T]
trait Quux[-T] extends Z[T]
class A
class B extends A
```
Given the above, what is the relative specificity of `Quux[B]` and `Z[A]`? Both Dotty and this implementation say that neither is as specific as the other even though `Z[B]` is as specific as `Z[A]` and `Quux[B]` is a subtype of `Z[B]`. @smarter what are your thoughts on this?